infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java

```diff
@@ -127,7 +142,45 @@ void prepare(boolean canCreate) throws InfrastructureException {
         throw new InfrastructureException(
             format("Creating the namespace '%s' is not allowed, yet it was not found.", name));
       }
-      create(name, client);
+      namespace = create(name, client);
+    }
+    label(namespace, labels);
+  }
+
+  /**
+   * Applies given `ensureLabels` into given `namespace` and update the `namespace` in the
+   * Kubernetes.
+   *
+   * <p>If we do not have permissions to do so (code=403), this method does not throw any exception.
+   *
+   * @param namespace namespace to label
+   * @param ensureLabels these labels should be applied on given `namespace`
+   * @throws InfrastructureException if something goes wrong with update, except lack of permissions
+   */
+  protected void label(Namespace namespace, Map<String, String> ensureLabels)
+      throws InfrastructureException {
+    Map<String, String> currentLabels = namespace.getMetadata().getLabels();
+    Map<String, String> newLabels =
+        currentLabels != null ? new HashMap<>(currentLabels) : new HashMap<>();
+
+    if (newLabels.entrySet().containsAll(ensureLabels.entrySet())) {
+      LOG.debug(
+          "Nothing to do, namespace [{}] already has all required labels.",
+          namespace.getMetadata().getName());
+      return;
+    }
+
+    newLabels.putAll(ensureLabels);
+    namespace.getMetadata().setLabels(newLabels);
+
+    try {
+      cheClientFactory.create().namespaces().createOrReplace(namespace);
```

Comment:
I have some doubts that we need to update labels for the namespace where che server is running. Maybe it's the job of the operator or another deployment tool.

Reply:
it is updating the namespace passed to this method -> `createOrReplace(namespace)`. The `cheClientFactory` is just creating the kubernetes client with `che` SA.
I should probably name the variables better or comment, so it's more clear what this is doing.